### ISSUE TITLE
Mario 64 filter/texture issues.

### ISSUE THREAD

**Issue by PimpUigi** [gonetz/GLideN64#2346]
I'm using GlideN64 rev.1111a76 (but also have GlideN64 4.0 which I switched back and forth to for this test)

The HUD in Super Mario 64 (as well as menu screens, text, 2D elements) seems to be completely unaffected by picking the Bilinear Option.
The HUD on GlindeN64 Bilinear looks exactly like Glide64 Point Sampled, and there is no difference when I switch back and forth. I can see the quality in the floor textures change, so I know bilinear is on and helping there....
When I switched to render in Original N64 resolution, even the pixels on Mario's face in the middle of the screen just appear clearly unfiltered, not like Bilinear would do...

For comparisons here I have included Glide64 Final Edition, where Bilinear changes the HUD entirely.
Furthermore I notice when switching to Point Sampled on Glide64 the 3D textures like the floor and walls are now unfiltered completely. Some people seem to prefer this (not me) and I'm wondering why switching between these options in GlideN64 don't seem to have the effects one would expect?

Glide64 Bilinear:
![image](https://user-images.githubusercontent.com/7344271/94829379-74a93e80-03d8-11eb-99ab-d97b5d2432d8.png)

Glide64 Point Sampled:
![image](https://user-images.githubusercontent.com/7344271/94829470-8b4f9580-03d8-11eb-9a2f-5bb9b490db41.png)

GlideN64 Bilinear:
![image](https://user-images.githubusercontent.com/7344271/94828415-6f97bf80-03d7-11eb-8ba6-9e7170108d73.png)

GlideN64 Point Sampled:
![image](https://user-images.githubusercontent.com/7344271/94828738-c7362b00-03d7-11eb-9353-f54db6acad5e.png)

GlideN64 Original N64 Resolution:
![image](https://user-images.githubusercontent.com/7344271/94828548-92c26f00-03d7-11eb-853b-4ed2da429a7c.png)

Super Mario 64's ending doesn't render correctly in any version of Glide. The texture is all broken up:
![image](https://user-images.githubusercontent.com/7344271/94828224-40814e00-03d7-11eb-9f83-20029ee168b8.png)

**Comment by gonetz** [comment]
GLideN64 has no "bilinear" and "point-sampling" mode.
The plugin filters texture only when filtering is enabled by game code. That is the texture filtering is controlled only by game, not by user. Glide64 has "force biliner" and "force point-sampled" modes, GLideN64 has only "Automatic". So, I don't understand what "GlideN64 Bilinear" and "GlideN64 Point Sampled" mean.
However I see the point: the HUD in SM64 indeed looks unfiltered. Filtering is set for HUD, so it should be filtered. I'll check what is wrong. Regarding ending screen - could you give me a savestate for it?

**Comment by PimpUigi** [comment]
[SM64 Ending.pj.zip](https://github.com/gonetz/GLideN64/files/5318266/SM64.Ending.pj.zip)
Is it difficult to implement Force Bilinear and Force Point Sampled?

**Comment by oddMLan** [comment]
> However I see the point: the HUD in SM64 indeed looks unfiltered. Filtering is set for HUD, so it should be filtered

Isn't it because the textrect has the COPY_MODE flag set, which always ignores the filtering setting? Even in angrylion the textures are unfiltered.
I agree the simplest solution would just giving the user the option to force the selected filter (either 3-point or bilinear) in all textures.

**Comment by gonetz** [comment]
> Isn't it because the textrect has the COPY_MODE flag set, which always ignores the filtering setting?

Yes, it is the reason. The HUD is not intended to be filtered. The plugin works correct here.

**Comment by gonetz** [comment]
> Is it difficult to implement Force Bilinear and Force Point Sampled?

It is easy, but I dislike that feature as potential source of issues and misunderstandings. 

Thanks for the ending savestate! It looks like this revealed a problem in texture loading code.

**Comment by PimpUigi** [comment]
OK, well on the other hand, it doesn't look like that on a TV...the lives, X, and Star aren't all pixelated. It looks like how the Force Bilinear mode makes it look.
So I understand (I think) that having the option in the GUI would cause a source of misunderstandings from the general users? What if the option was simply hidden in the .ini file so only people hunting for it specifically will be able to find it?

**Comment by oddMLan** [comment]
That's because a CRT display tends to smooth out the whole image. You could accomplish the same effect with a CRT shader in RetroArch. (Or with GLideN64 and ReShade, probably).

**Comment by gonetz** [comment]
Hmm, the wrong ending screen is probably a core issue. I see the same issue with angrylion's plugin, with PJ64 and mupen64plus.

**Comment by PimpUigi** [comment]
Wouldn't reshade shade shade the whole image at that point though? (which would be bad)
I would just use Glide64 but it can't enable "Try to adjust game to fit" widescreen like GlideN64 can. And in my latest compile of my custom Mario 64 I have the clipping fixed with the following code.

```
halfFov = (gCurGraphNodeCamFrustum->fov / 2.0f + 1.0f) * 32768.0f / 180.0f + 0.5f;
    hScreenEdge = -matrix[3][2] * sins(halfFov) / coss(halfFov);
    hScreenEdge *= 1.5f;
```

And yeah I can only get the ending screen to look right with Jabo.

**Comment by gonetz** [comment]
> What if the option was simply hidden in the .ini file so only people hunting for it specifically will be able to find it?

No, until texturing code is not 100% correct I prefer to have as least sources of potential errors as possible.
This code is already nightmare to debug.

**Comment by PimpUigi** [comment]
Well I can definitely respect that reasoning, needing to know your texturing code is 100% correct before red herring functions are out there.
So, how can I go about making the change for my own compile? Do I just look for that textrect Copy_mode setting and adjust it?

**Comment by gonetz** [comment]
> And yeah I can only get the ending screen to look right with Jabo.

Yep, it looks right with Jabo's plugin. So, it is not a core issue, unless Jabo made some hack for that case.
It is hard to believe that angrylion's plugin has a bug in texture loading code.

> So, how can I go about making the change for my own compile? Do I just look for that textrect Copy_mode setting and adjust it?

You may just set uTextureFilterMode to 1 in 
src\Graphics\OpenGLContext\GLSL\glsl_CombinerProgramUniformFactory.cpp
It will force texture filtering.
Set uTextureFilterMode to 0 to completely disable filtering.

**Comment by PimpUigi** [comment]
Perhaps Angrylion isn't aware of the ending screen bug either. I appreciate your help, your time, I'm certainly in no rush to fix the bug, just _thank you_ for showing interest in fixing it and for the advice/help for my filtering preference. (like with Goldeneye I find everything to be just right, but that's probably because it has a high res HUD pack)

I hope this time next year or something we'll have found a big breakthrough that takes care of it, but I know the reality of how these things can go. Respect. : )

**Comment by oddMLan** [comment]
> Yep, it looks right with Jabo's plugin. So, it is not a core issue, unless Jabo made some hack for that case.
> It is hard to believe that angrylion's plugin has a bug in texture loading code.

How do GLideN64 and Angrylion look with Mupen64Plus?

**Comment by gonetz** [comment]
> How do GLideN64 and Angrylion look with Mupen64Plus?

The same issue as with Project64.

**Comment by PimpUigi** [comment]
I sent you an email Gonetz. Just hoping it made it over to you.

**Comment by PimpUigi** [comment]
I set 
int textureFilter = 1;

But it resulted in this strange instance of filtering on text and menus, but _not_ the HUD. 
![image](https://user-images.githubusercontent.com/7344271/95405225-3e425680-08e5-11eb-8571-92607fe08def.png)

![image](https://user-images.githubusercontent.com/7344271/95405245-4d290900-08e5-11eb-873b-cfe4684519a4.png)
I'm also unsure of what I should change to prevent the filtered textures there from seemingly, tiling themselves for a few pixels??
Glide64 doesn't do that.

**Comment by gonetz** [comment]
Yes, I forgot that copy mode uses different shader with no filtering. So, "force filtering" is not that easy to implement.

**Comment by PimpUigi** [comment]
Ack. At this point it almost seems easier to find a way to change Glide64's aspect ratio code and adjust the "16:9" setting to a "Try to adjust game to fit" setting.
That's really what Glide64 lacks is the robust Widescreen ability of GlideN64.

**Comment by PimpUigi** [comment]
Had an idea while I was playing the game with Angrylion. Since it has a Linear interpolation setting...
What if we had a checkbox in the Emulation tab that could render copy_mode / textrect elements at Original N64 Resolution with linear interpolation like Angrylion??

When I was setting up Goldeneye with 1964 about a year ago, the HUD's texture pack didn't have the ammo types textured yet. So they were all pixelated and I couldn't get them interpolated to look like the original hardware. I kept flipping back between my monitors composite input with my N64 hooked up, and the PC running 1964 in full screen. And the ammo / bullet graphic would look smooth on the N64, but pixelated in 1964.
Now I just walked away for a year and came back to a more complete Texture Pack.

Testing on Retroarch now so I can apply a simple interpolation filter to 320x240 GlideN64 mupen...(I will post pictures if it helps, but I thought I'd just keep this post clean)
The accurate way to display HUD's / copy_mode / textrects seems to definitely be to render at 320x240 with linear interpolation.

It looks correct in both Angrylion on PJ64 and GlideN64 on Retroarch with the basic bilinear shader.
Is this what "Render 2D Elements in N64 resolution" is about? Just with Nearest Neighbor instead of Linear??? :o

**Comment by ToruTheRedFox** [comment]
AngryLion's bilinear mode is only for the upscaling of the final image to different resolutions

it doesn't change anything with the actual rendering

**Comment by PimpUigi** [comment]
How can I put this another way???

These two pictures are taken from Angrylion...with Linear filter on. They look _virtually the same_ on my 50 inch Kuro Plasma TV as the N64 connected to the TV is (switching between the PC emulator in fullscreen to the real hardware)
![image](https://user-images.githubusercontent.com/7344271/95661363-a3bc6000-0afc-11eb-94eb-66e226bcc740.png)
![image](https://user-images.githubusercontent.com/7344271/95661338-753e8500-0afc-11eb-8d88-45348b3940ce.png)

Nearest Neighbor Angrylion is _not_ what the actual hardware looks like. Looking at both Mario's face, and the HUD.
![image](https://user-images.githubusercontent.com/7344271/95661382-e3834780-0afc-11eb-962b-b6e55c0ed086.png)

**Comment by ToruTheRedFox** [comment]
TVs themselves do linear filtering on inputs afaik

**Comment by PimpUigi** [comment]
On NES/SNES/and Sega Genesis emulators, that's considered a factor in accuracy and so they all offer a "bilinear filtered" mode. Separate from the over-shaded, often scalines ridden "TV Mode" or "NTSC Modes"

As a dev, we design sprites with that fully in mind. There are also lots of YouTube videos that explain little technical things devs took advantage of on old consoles like this https://www.youtube.com/watch?v=x0weL5XDpPs

So I mean, I hope Gonetz at least feels that both ways are accurate emulation. : (
I used to prefer nearest neighbor, and then I started being a dev and things changed. Now I prefer the accuracy.

**Comment by PimpUigi** [comment]
Strange. It looks correct when "Render 2D Elements in N64 Resolution" is checked _plus_  HLE Graphics are **disabled** (on Project64) - does this mean the issue with the HLE graphics???????
There are lots of graphical glitches with HLE unchecked though. Like _lots and lots._

![image](https://user-images.githubusercontent.com/7344271/95665029-944c0f80-0b1a-11eb-9879-4807fa7b1e58.png)

**Comment by oddMLan** [comment]
If you use a UltraHDMI the output is more like the nearest neighbor interpolation. It's the first time that I hear someone liking the inherent N64 DAC blurriness.

**Comment by PimpUigi** [comment]
Every single current solution to fix the HUD leaves the game looking like crap...I definitely never liked how the N64 was far blurrier and foggier than PSX or Dreamcast. It is really awesome to have the levels, models, and textures rendered in 1920x1080 like GlideN64 does. However I'm sure you still prefer the Bilinear filtering on those models and textures over the point sampling that Glide64 can do. See the floor and walls in the castle of the Point Sample image with Bilinear filtering disabled??? Horrible. - So I think we agree, unless you like the below image, that Bilinear filtering has its place. 
https://user-images.githubusercontent.com/7344271/94829470-8b4f9580-03d8-11eb-9a2f-5bb9b490db41.png
There's just no need to apply an _extra_ blur over that, I'd certainly agree anything that would shade the entire image to the point that it got the HUD looking right at that point, would blur the rest of the image so bad that you'd puke. : (

Why _can't_ we have a checkbox option (or dropdown option? like a fourth option under "Optimized" "Not Optimized" "Disabled" "Interpolated") in the "Render 2D Elements in N64 Resolution" area ... 
It would then look like this:
https://user-images.githubusercontent.com/7344271/95665029-944c0f80-0b1a-11eb-9879-4807fa7b1e58.png
And that honestly looks better than the Jabo / Glide64 "force" filter options _by far and away_.

**Comment by ToruTheRedFox** [comment]
it'd look really bad with accurate 3-point filtering

**Comment by PimpUigi** [comment]
I'm sorry, I'm just not following the logic of that post. The screenshot I last posted was with GlideN64 - I simply manipulated the compile + had to disable HLE in PJ64 - this caused a whole handful of other rendering issues elsewhere in the game, but _that is what it would look like._

Does that screenshot right above your post "look really bad" ???

**Comment by ToruTheRedFox** [comment]
the N64 does bilinear filtering in a different way to PC hardware

if you want to have the output to look like it does on N64, you should enable the 3-point filtering option, and what I'm saying is that due to the nature of that (pretty hacky) type of bilinear, the HUD would just look like it just got smudged diagonally

**Comment by PimpUigi** [comment]
I've now reprogrammed a chunk of the HUD to display in G_TF_BILERP instead of TF_POINT
![image](https://im3.ezgif.com/tmp/ezgif-3-f9a4d8436e3e.gif)

But naturally when I reprogrammed the Main HUD away from G_CYC_COPY I ran into problems...
![image](https://user-images.githubusercontent.com/7344271/96092738-45470700-0e99-11eb-86ed-bc5a72a63fd7.png)
(first it duplicated each sprite x4, but then when I got it down to 1, I couldn't really resize those sprites without editing the complicated math in Print.c)

You're sure there's not something I can't write in the uniforn factory like 
int textureFilter = gDP.otherMode.textureFilter;
if {
		if (gDP.otherMode.cycleType < G_CYC_COPY) {
int textureFilter = 1;}

**Comment by PimpUigi** [comment]
(fixed power meter image link)

@gonetz does something like this seem feasible?
int textureFilter = gDP.otherMode.textureFilter;
if {
if (gDP.otherMode.cycleType < G_CYC_COPY) {
int textureFilter = 1;}

The way the code works in game, I don't have to apply BILERP to the text boxes, which leaves all the text still looking crisp instead of faded. So really with the reprogramming I've done, I only need to force filtering onto the G_CYC_COPY instances (which is literally _only_ the Main Hud & Press Start glyph)

**Comment by oddMLan** [comment]
[GLideN64-wtl-force_filter_copymode.zip](https://github.com/gonetz/GLideN64/files/5394931/GLideN64-wtl-force_filter_copymode.zip)
Attached is a dll (for PJ64 only) that forces filtering for COPY_MODE texrects (diff included) with whichever filter you select

However it doesn't look good and the filtering isn't applied regular texrects (non copy mode) that don't explicitly request being filtered (like the POW meter). 

![GLideN64_SUPER_MARIO_64_004](https://user-images.githubusercontent.com/7278372/96327442-79800c00-0fee-11eb-83cb-15c99052e7f4.png)

This makes me believe that the devs never intended for the textures to be filtered in the HUD, and if you look at the beta SM64 screenshots they use a different style that's clearly pixelated.

![sm64_beta_recreation](https://user-images.githubusercontent.com/15708278/95883743-2fd7bd00-0d73-11eb-90a7-63ad19f13d49.png)

So this isn't really a GLideN64 issue, since it's just following the filtering that the game mandates.
What the N64 does is smear the whole picture before sending it to the TV to get the pixels in the HUD look "smoother" (at the cost of making everything else blurry). Like you yourself said, applying a screen filter thru Reshade so that the HUD looks like you want would blur the rest of the picture as well. There isn't a good way for GLideN64 to know what you want to look filtered except through on a case-by-case basis with ugly hacks, which go outside of the scope of this project. And the time of @gonetz is very limited atm, plus he has tons of bugs to fix first.

If you have issues in LLE with depth disable N64 depth compare and use FXAA (what the N64 uses to smooth out the jagged edges) instead of MSAA.

**Comment by PimpUigi** [comment]
I certainly respect all that context, and I respect Gonetz is very limited at the moment. I hope whatever burdens he is under pass, as he has been very respectful and kind to me throughout. I greatly appreciate you taking the time to hook me up with a DLL like that! 
Thank you oddLMan. : )

I also did put the work in to be able to compile in the future if I need to.
And yeah I see what you mean about it not looking right, and things marked G_CYC_COPY aren't responding to the G_TX_CLAMP command that prevents them from repeating like they are on the right and bottom of the 4 in your screen shot. So unfortunately I have only reached the next hurdle. : (
As it seems with every seemingly final solution to this paltry dilema...

Like when I set the File Select screen to have TF_BILERP, the Stars have that same "tiled" effect where you can see them repeat to the right and bottom by a pixel or two. Setting the G_TX_CLAMP flag instead of WRAP fixes that; but G_CYC_COPY _ignores that_ just like with the G_TF_BILERP command. I wonder why? 
I can change the G_TX_NOMIRROR to G_TX_MIRROR for now, which makes them tile in a flip flop / mirrored way, effectively extending the bottom and right pixels down and to the right (and it seems like it's only by 1 pixel in fact)

In Glide64 they got around this somehow, so do you think there are some relevant numbers in the .cpp files I can toy around with changing numbers on to find the solution? (once I know where to adjust code I'm sure I can make progress)

**Comment by PimpUigi** [comment]
Looking through Glide64's code, it does look like they are having an extensive bit of code about detecting forced wrapping in texrect instances and changing it to clamping. (at least it looks like it to me)

```
            {
                gfxTextureFilterMode_t filter = (g_settings->filtering() == CSettings::**Filter_ForceBilinear**) ? GFX_TEXTUREFILTER_BILINEAR : GFX_TEXTUREFILTER_POINT_SAMPLED;
                gfxTexFilterMode(tmu, filter, filter);
            }

            if (rdp.cur_cache[i])
            {
                gfxTextureClampMode_t mode_s, mode_t;
                int clamp_s, clamp_t;
                **if (rdp.force_wrap && !rdp.texrecting)
                {
                    clamp_s = rdp.tiles(tile).clamp_s && rdp.tiles(tile).lr_s - rdp.tiles(tile).ul_s < 256;
                    clamp_t = rdp.tiles(tile).clamp_t && rdp.tiles(tile).lr_t - rdp.tiles(tile).ul_t < 256;
                }**
                else
                {
                    clamp_s = (rdp.tiles(tile).clamp_s || rdp.tiles(tile).mask_s == 0) &&
                        rdp.tiles(tile).lr_s - rdp.tiles(tile).ul_s < 256;
                    clamp_t = (rdp.tiles(tile).clamp_t || rdp.tiles(tile).mask_t == 0) &&
                        rdp.tiles(tile).lr_t - rdp.tiles(tile).ul_t < 256;
                }

                if (rdp.cur_cache[i]->f_mirror_s)
                    mode_s = GFX_TEXTURECLAMP_MIRROR_EXT;
                else if (rdp.cur_cache[i]->f_wrap_s)
                    mode_s = GFX_TEXTURECLAMP_WRAP;
                else if (clamp_s)
                    mode_s = GFX_TEXTURECLAMP_CLAMP;
                else
                {
                    if (rdp.tiles(tile).mirror_s && !g_settings->hacks(CSettings::hack_Zelda))
                        mode_s = GFX_TEXTURECLAMP_MIRROR_EXT;
                    else
                        mode_s = GFX_TEXTURECLAMP_WRAP;
                }

                if (rdp.cur_cache[i]->f_mirror_t)
                    mode_t = GFX_TEXTURECLAMP_MIRROR_EXT;
                else if (rdp.cur_cache[i]->f_wrap_t)
                    mode_t = GFX_TEXTURECLAMP_WRAP;
                else if (clamp_t)
                    mode_t = GFX_TEXTURECLAMP_CLAMP;
                else
                {
                    if (rdp.tiles(tile).mirror_t && !g_settings->hacks(CSettings::hack_Zelda))
                        mode_t = GFX_TEXTURECLAMP_MIRROR_EXT;
                    else
                        mode_t = GFX_TEXTURECLAMP_WRAP;
                }

                gfxTexClampMode(tmu,
                    mode_s,
                    mode_t);
            }
```

**Comment by PimpUigi** [comment]
I haven't been able to make any progress on my own the last couple months. Kim do you think it'd be possible to edit the filtering for COPY_MODE texrects to force texture clamping?

**Comment by oddMLan** [comment]
> Kim do you think it'd be possible to edit the filtering for COPY_MODE texrects to force texture clamping?

I wouldn't know how to do that myself.

**Comment by PimpUigi** [comment]
I just wanted to post an update:
I was setting up an install of Project 64 for someone else so they could play my custom _Luigi in Super Mario 64_ and they had Glide64 set to Automatic, and while I always had mine set to Force_Bilinear I never noticed...
Glide64 is respecting the code I programmed.

When I program segment2.c with
`// 0x0200EC60 - 0x0200EC98
const Gfx dl_hud_img_begin[] = {
    gsDPPipeSync(),
    gsDPSetCycleType(G_CYC_COPY),
    gsDPSetTexturePersp(G_TP_NONE),
    gsDPSetAlphaCompare(G_AC_THRESHOLD),
    gsDPSetEnvColor(255, 255, 255, 255),
    gsDPSetTextureFilter(G_TF_BILERP),
    gsDPSetTextureDetail(G_TD_CLAMP),
    gsSPEndDisplayList(),
};

// 0x0200EC98 - 0x0200ECC8
const Gfx dl_hud_img_load_tex_block[] = {
    gsDPTileSync(),
    gsDPSetTile(G_IM_FMT_RGBA, G_IM_SIZ_16b, 0, 0, G_TX_LOADTILE, 0, G_TX_CLAMP, 4, G_TX_NOLOD, G_TX_CLAMP, 4, G_TX_NOLOD),
    gsDPLoadSync(),
    gsDPLoadBlock(G_TX_LOADTILE, 0, 0, 16 * 16 - 1, CALC_DXT(16, G_IM_SIZ_16b_BYTES)),
    gsDPSetTile(G_IM_FMT_RGBA, G_IM_SIZ_16b, 4, 0, G_TX_RENDERTILE, 0, G_TX_CLAMP, 4, G_TX_NOLOD, G_TX_CLAMP, 4, G_TX_NOLOD),
    gsDPSetTileSize(0, 0, 0, (16 - 1) << G_TEXTURE_IMAGE_FRAC, (16 - 1) << G_TEXTURE_IMAGE_FRAC),
    gsSPEndDisplayList(),
};`

I get this:
![image](https://user-images.githubusercontent.com/7344271/103036645-3b482100-4538-11eb-8c87-eaf7b0f7c8a1.png)
Perfectly unfiltered Text, while the HUD respects both the 
    `gsDPSetTextureFilter(G_TF_BILERP) `as well as the corresponding `gsDPSetTile(G_IM_FMT_RGBA, G_IM_SIZ_16b, 4, 0, G_TX_RENDERTILE, 0, G_TX_CLAMP, 4, G_TX_NOLOD, G_TX_CLAMP, 4, G_TX_NOLOD),`

Now if I set the default code, it looks like this:
![image](https://user-images.githubusercontent.com/7344271/103037060-1ef8b400-4539-11eb-98c4-3da5bfb697ea.png)
The default code entails `gsDPSetTextureFilter(G_TF_POINT)` and `gsDPSetTile(G_IM_FMT_RGBA, G_IM_SIZ_16b, 4, 0, G_TX_RENDERTILE, 0, G_TX_WRAP | G_TX_NOMIRROR, 4, G_TX_NOLOD, G_TX_WRAP | G_TX_NOMIRROR, 4, G_TX_NOLOD)`

So I really think it's on GlideN64 to respect the `G_TX_CLAMP` command and ` gsDPSetTextureFilter(G_TF_BILERP) `when it comes to `gsDPSetCycleType(G_CYC_COPY)`

I'd roll with Glide64 except while I was editing these .c files, I am able to fix SM64's draw distance to work with GlideN64's very excellent widescreen mode "Try to adjust game to fit", and we get this:
![image](https://user-images.githubusercontent.com/7344271/103037653-5ae04900-453a-11eb-93ad-4cbd5eb68df4.png)
Not a single object gets clipped at the edges of a 16:9 resolution now.

Now it took me quite some time to program in these fixes, learn the .c code to add the SMB1, 3, and Classic Luigi colors, as well as to learn the .c code to filter the Power Meter and HUD objects. So I am happy to be patient because I understand the effort involved in getting the code down, but I just really need you guys' help to get the last bit of the way with GlideN64.
GlideN64 is the gold standard for a reason and it's because you guys work so hard to be that Excellent bar of quality. So I hope it can be updated to respect `G_TX_CLAMP` and  `gsDPSetTextureFilter(G_TF_BILERP)`, and that's so much better than the "Force Bilinear" mode I was trying to do at first because it leaves all the text readable.

I can provide you with my ROM mods if it helps you test.
Thank you for taking the time and effort that you have, I'm also willing to toss over donations for the time you have taken to help me.

**Comment by PimpUigi** [comment]
![image](https://user-images.githubusercontent.com/7344271/103041638-f4acf380-4544-11eb-95c6-046d78a16c92.png)
Here's Glide64 with `gsDPSetTextureFilter(G_TF_BILERP)` set but `G_TX_WRAP` instead of clamp...so I guess Glide64 either auto clamps and ignores the Wrap command, or some other kind of trick I have yet to figure out. : (

Maybe it is adding an extra line of alpha pixels on all four sides before a Wrap. I don't know.

**Comment by standard-two-simplex** [comment]
> So I really think it's on GlideN64 to respect the G_TX_CLAMP command and gsDPSetTextureFilter(G_TF_BILERP)when it comes to gsDPSetCycleType(G_CYC_COPY)

Copy mode is meant to copy a texture to screen very fast. Instead of reading 4 neighbour texels and filtering them into one (like in one cycle mode), it reads four consecutive texels and copies the to the screen. As a result, 4 texels are rendered per cycle instead of just one.

However, the RDP is not able to keep up with most of its pipeline at this fillrate. I can't perform enough operations to shade 4 pixels in one cycle. Therefore a lot of features are disabled in copy mode. Texture coordinate clamping and texture filtering are among those disabled.

I would say that GlideN64 respects the behaviour of the N64 in this regard.

If you want to hack the rom to render those rectangles with bilinear filtering and clamping you need to do the following.
- Set cycle type to 1 cycle.
- Set filtering mode to bilerp.
- Change the texrect command to use dsdx=1 instead of 4.
- Set the combiner to output the same value as the input.
- Disable the blender.

If you want to hack GlideN64 to force bilinear filtering in copy mode (which I think is a bad idea) you need to (assuming you want it for standard GL),
- Modify `READ_TEX()` in `ShaderFragmentReadTexCopyMode()` in `glsl_CombinerProgramBuilder.cpp` and change `"    name = texelFetch(tex, ivec2(tcData[0]),0);					\\\n"` for code present int `TEX_FILTER()`.

Again, I believe this is not a good idea as it may break other games that rely on correct behaviour of copy mode.

**Comment by standard-two-simplex** [comment]
> But naturally when I reprogrammed the Main HUD away from G_CYC_COPY I ran into problems...
.....
(first it duplicated each sprite x4, but then when I got it down to 1, I couldn't really resize those sprites without editing the complicated math in Print.c)

That is the closest you got to accomplishing the first path, you only need to change the texrect command that renders the hud.

**Comment by PimpUigi** [comment]
Not as pretty as forcing GlideN64 to do it (but again, that method left the Textures repeating FWIW)
But that Texrect changing 4 << 10 to 1 << 10 did the trick!
In the same ()'s there were 15's I had to change to 16's to get the full glyph to display, but you solved it!
The texrect's were hidden in HUD.c (camera Glyphs) and Print.c (Life and Star glyphs) while the main code I had been editing was in segment2.c

So yeah, a lot of figuring things out!
Thank you.
![image](https://user-images.githubusercontent.com/7344271/103083921-a67f0b00-45ab-11eb-94f4-114c346dee39.png)

vs.

![image](https://user-images.githubusercontent.com/7344271/103084531-3a050b80-45ad-11eb-8197-fdc62644891f.png)
Widescreen is worth the extra jaggies and slightly blurrier filtering.
But that's kind of a shame it's lower quality filtering with the game hack.

I am deeply appreciative all the same, what a breakthrough!

**Comment by PimpUigi** [comment]
Just wondering, since you seem to be very knowledgeable about N64 dev stuff. since the Life Icon was initially cut off by one pixel on the bottom until I changed the 15 to a 16, could I have moved the icon up one pixel so the top pixel would be cut off instead of the bottom one???

**Comment by standard-two-simplex** [comment]
> could I have moved the icon up one pixel so the top pixel would be cut off instead of the bottom one???

No, copy mode rasters the primitive differently, so texrect width and height are decreased by one usually to compensate for it. If you switch to 1cycle mode you should add 1 to width and height to render it the same size.

**Comment by PimpUigi** [comment]
Yeah I figured all that out (as shown in the screenshots) so there's not an option to move the glyph up a pixel or two if I wanted the top to be cut off. But I can cut the bottom off just fine if I wanted. Gotcha.
Truly appreciative to you guys, really I would not have pulled this off without the help here and I think you guys all kick ass.

**Comment by ToruTheRedFox** [comment]
you can probably cut the top off by messing with the UVs of the triangles the game generates to actually draw the glyphs
